x-pack/plugins/maps/public/connected_components/map_settings_panel/custom_icons_panel.tsx

```diff
@@ -0,0 +1,202 @@
+ * 2.0.
+ */
+
+import React, { Component, Fragment } from 'react';
+import {
+  EuiButtonEmpty,
+  EuiListGroup,
+  EuiPanel,
+  EuiSpacer,
+  EuiText,
+  EuiTextAlign,
+  EuiTextColor,
+  EuiTitle,
+} from '@elastic/eui';
+import { i18n } from '@kbn/i18n';
+import { FormattedMessage } from '@kbn/i18n-react';
+import { DEFAULT_CUSTOM_ICON_CUTOFF, DEFAULT_CUSTOM_ICON_RADIUS } from '../../../common/constants';
+import { getIsDarkMode } from '../../kibana_services';
+// @ts-expect-error
+import { getCustomIconId } from '../../classes/styles/vector/symbol_utils';
+import { SymbolIcon } from '../../classes/styles/vector/components/legend/symbol_icon';
+import { CustomIconModal } from '../../classes/styles/vector/components/symbol/custom_icon_modal';
+import { CustomIcon } from '../../../common/descriptor_types';
+
+interface Props {
+  customIcons: CustomIcon[];
+  updateCustomIcons: (customIcons: CustomIcon[]) => void;
+  deleteCustomIcon: (symbolId: string) => void;
+}
+
+interface State {
+  isModalVisible: boolean;
+  selectedIcon?: CustomIcon;
+}
+
+export class CustomIconsPanel extends Component<Props, State> {
+  public state = {
+    isModalVisible: false,
+    selectedIcon: undefined,
+  };
+
+  private _handleIconEdit = (icon: CustomIcon) => {
+    this.setState({ selectedIcon: icon, isModalVisible: true });
+  };
+
+  private _handleNewIcon = () => {
+    this.setState({ isModalVisible: true });
+  };
+
+  private _renderModal = () => {
+    if (!this.state.isModalVisible) {
+      return null;
+    }
+    if (this.state.selectedIcon) {
+      const { symbolId, label, svg, cutoff, radius } = this.state.selectedIcon;
+      return (
+        <CustomIconModal
+          title={i18n.translate('xpack.maps.mapSettingsPanel.editCustomIcon', {
+            defaultMessage: 'Edit custom icon',
+          })}
+          symbolId={symbolId}
+          label={label}
+          svg={svg}
+          cutoff={cutoff || DEFAULT_CUSTOM_ICON_CUTOFF}
+          radius={radius || DEFAULT_CUSTOM_ICON_RADIUS}
+          onSave={this._handleSave}
+          onCancel={this._hideModal}
+          onDelete={this._handleDelete}
+        />
+      );
+    }
+    return (
+      <CustomIconModal
+        title={i18n.translate('xpack.maps.mapSettingsPanel.addCustomIcon', {
+          defaultMessage: 'Add custom icon',
+        })}
+        cutoff={DEFAULT_CUSTOM_ICON_CUTOFF}
+        radius={DEFAULT_CUSTOM_ICON_RADIUS}
+        onSave={this._handleSave}
+        onCancel={this._hideModal}
+      />
+    );
+  };
+
+  private _hideModal = () => {
+    this.setState({ isModalVisible: false, selectedIcon: undefined });
+  };
+
+  private _handleSave = (icon: CustomIcon) => {
+    const { symbolId, label, svg, cutoff, radius } = icon;
+
+    const icons = [
+      ...this.props.customIcons.filter((i) => {
+        return i.symbolId !== symbolId;
+      }),
+      {
+        symbolId: symbolId || getCustomIconId(),
```

Comment:
CustomIcon defines symbolId as `symbolId: string;` so change `symbolId: symbolId || getCustomIconId(),` to just `symbolId,`